### DISCUSSION TITLE
Keycloak instance configured for HTTPS is redirecting localhost requests to the HTTP endpoint to HTTPS but on the HTTP port, resulting in errors

### DISCUSSION THREAD

**Discussion by DamianEdwards** [keycloak/keycloak#32319] [Q&A] [answered]
I have Keycloak running in a container (v24.0), configured for TLS. The HTTPS endpoints work as expected, but requests to the HTTP endpoints are being redirected to HTTPS but on the HTTP port, resulting in errors.

Keycloak is running locally in a container with the following port bindings:

```json
"PortBindings": {
  "8080/tcp": [
    {
      "HostIp": "127.0.0.1",
      "HostPort": "0"
    }
  ],
  "8443/tcp": [
    {
      "HostIp": "127.0.0.1",
      "HostPort": "0"
    }
  ]
}
```

The `8080` binding was assigned host port `62919` and the `8443` binding was assigned host port `62920`:

| Host Port | Container Port |
| ---- | ---- |
| 62919 | 8080 |
| 62920 | 8433 |

The following environment variables are set in order to enable TLS. The `QUARKUS_HTTP_HTTP2` is also set in order to disable HTTP/2 as a workaround for large headers issue (431 responses) as discussed in #13933 and #10236:

| Env var name | Value |
| --- | --- |
| KC_HTTPS_CERTIFICATE_FILE | /dev-certs/dev-cert.pem |
| KC_HTTPS_CERTIFICATE_KEY_FILE | /dev-certs/dev-cert.key |
| KC_HTTPS_PORT | 8443 |
| QUARKUS_HTTP_HTTP2 | false |

Requests behave as follows:

| Requested URL | Result |
| --- | --- |
| https://localhost:62920 | 307 redirects x2 to https://localhost:62920/admin/master/console |
| https://127.0.0.1:62920 | 307 redirects x2 to https://127.0.0.1:62920/ (cert error due to host name mismatch but not relevant) |
| http://localhost:62919 | 307 redirects to https://localhost:62919/ |
| http://127.0.0.1:62919 | 307 redirects x2 to http://127.0.0.1:62919/admin/master/console |

From the above, you can see that requests to **http**://localhost attempt to redirect to **https**://localhost but on the same port, which of course won't work. Requests to `127.0.0.1` do not redirect this way.

I tried setting the `KC_HOSTNAME` environment variable to `localhost` and the behavior was the same.

**Comment by DamianEdwards** [comment]
Here's a screencap (from a separate run so host ports don't match description above):

https://github.com/user-attachments/assets/aaf54a54-3a9e-4ab4-a6d2-acf0f021faf2

**Reply by shawkins** [reply]
You don't quite have a complete description of your setup - do you have http-enabled=true, is hostname-strict=false? Please provide all of your http and hostname settings.

**Reply by DamianEdwards** [reply]
I haven't changed any other settings so I'm assuming they're their default values. Here's what I think is the relevant log line from container startup:
```
[org.keycloak.quarkus.runtime.hostname.DefaultHostnameProvider] (main) Hostname settings: Base URL: <unset>, Hostname: <request>, Strict HTTPS: false, Path: <request>, Strict BackChannel: false, Admin URL: <unset>, Admin: <request>, Port: -1, Proxied: false
```

**Reply by shawkins** [reply]
What does the log say about what ports keycloak is listening on? If you don't have http-enabled=true (false is the default) then it shouldn't be using the 8080 port.

**Reply by DamianEdwards** [reply]
`[io.quarkus] (main) Keycloak 24.0.5 on JVM (powered by Quarkus 3.8.4) started in 3.195s. Listening on: http://0.0.0.0:8080/ and https://0.0.0.0:8443/`

**Reply by shawkins** [reply]
Please use show-config to provide the full configuration.

**Reply by DamianEdwards** [reply]
```
Current Mode: development
Current Configuration:
        kc.cache =  local (PersistedConfigSource)
        kc.config.built =  true (SysPropConfigSource)
        kc.db =  dev-file (PersistedConfigSource)
        kc.hostname-strict =  false (PropertiesConfigSource[source=jar:file:///opt/keycloak/lib/lib/main/org.keycloak.keycloak-quarkus-server-24.0.5.jar!/META-INF/keycloak.conf])
        kc.hostname-strict-https =  false (PropertiesConfigSource[source=jar:file:///opt/keycloak/lib/lib/main/org.keycloak.keycloak-quarkus-server-24.0.5.jar!/META-INF/keycloak.conf])
        kc.http-enabled =  true (PropertiesConfigSource[source=jar:file:///opt/keycloak/lib/lib/main/org.keycloak.keycloak-quarkus-server-24.0.5.jar!/META-INF/keycloak.conf])
        kc.https-certificate-file =  /dev-certs/dev-cert.pem (KcEnvVarConfigSource)
        kc.https-certificate-key-file =  /dev-certs/dev-cert.key (KcEnvVarConfigSource)
        kc.https-port =  8443 (KcEnvVarConfigSource)
        kc.https.certificate.file =  /dev-certs/dev-cert.pem (KcEnvVarConfigSource)
        kc.https.certificate.key.file =  /dev-certs/dev-cert.key (KcEnvVarConfigSource)
        kc.https.port =  8443 (KcEnvVarConfigSource)
        kc.log-console-output =  default (PropertiesConfigSource[source=jar:file:///opt/keycloak/lib/lib/main/org.keycloak.keycloak-quarkus-server-24.0.5.jar!/META-INF/keycloak.conf])
        kc.log-file =  ${kc.home.dir:default}${file.separator}data${file.separator}log${file.separator}keycloak.log (PropertiesConfigSource[source=jar:file:///opt/keycloak/lib/lib/main/org.keycloak.keycloak-quarkus-server-24.0.5.jar!/META-INF/keycloak.conf])
        kc.run.in.container =  true (KcEnvVarConfigSource)
        kc.spi-theme-cache-templates =  false (PropertiesConfigSource[source=jar:file:///opt/keycloak/lib/lib/main/org.keycloak.keycloak-quarkus-server-24.0.5.jar!/META-INF/keycloak.conf])
        kc.spi-theme-cache-themes =  false (PropertiesConfigSource[source=jar:file:///opt/keycloak/lib/lib/main/org.keycloak.keycloak-quarkus-server-24.0.5.jar!/META-INF/keycloak.conf])
        kc.spi-theme-static-max-age =  -1 (PropertiesConfigSource[source=jar:file:///opt/keycloak/lib/lib/main/org.keycloak.keycloak-quarkus-server-24.0.5.jar!/META-INF/keycloak.conf])
        kc.version =  24.0.5 (SysPropConfigSource)
```

**Reply by shawkins** [reply]
Ok, so you are running in the dev profile so http-enabled=true and hostname-strict=false as well. 

> From the above, you can see that requests to http://localhost attempt to redirect to https://localhost but on the same port, which of course won't work. Requests to 127.0.0.1 do not redirect this way.

For completeness, are you missing an entry in that table - one that shows http://localost:62919/ using a 307 redirecting to https://localhost:62919/admin/master/console

Keycloak does not directly issue 307 redirects. It is likely the behavior of chrome trying to see if https is available: https://stackoverflow.com/questions/70705778/chrome-redirects-and-makes-2-requests-of-the-same-page-307-internal-redirect

If you check the headers directly being set by the keycloak response, it is not setting anything that would given an indication that https should be used.

**Reply by DamianEdwards** [reply, marked as answer]
> For completeness, are you missing an entry in that table - one that shows http://localost:62919/ using a 307 redirecting to https://localhost:62919/admin/master/console

Actually I think I just have a few copy-paste errors. I'll update it.

> Keycloak does not directly issue 307 redirects. It is likely the behavior of chrome trying to see if https is available

That's very interesting. That sounds like HSTS and indeed I can see that's what the reason cited is:

![image](https://github.com/user-attachments/assets/36a6f9df-b525-4fe6-8d06-a0899d9c0ea7)

I can confirm that after deleting the HSTS policy for localhost from my browser the issue no longer reproduces. Thanks for the insights!

**Reply by DamianEdwards** [reply]
@shawkins looking further at the headers coming back from the initial request to Keycloak, it seems like it's actually sending the HSTS header? See the `Strict-Transport-Security: max-age=31536000; includeSubDomains` response header in the curl trace below:

```shell
curl https://localhost:59569 -v
* Host localhost:59569 was resolved.
* IPv6: ::1
* IPv4: 127.0.0.1
*   Trying [::1]:59569...
*   Trying 127.0.0.1:59569...
* Connected to localhost (127.0.0.1) port 59569
* schannel: disabled automatic use of client certificate
* ALPN: curl offers http/1.1
* ALPN: server did not agree on a protocol. Uses default.
* using HTTP/1.x
> GET / HTTP/1.1
> Host: localhost:59569
> User-Agent: curl/8.8.0
> Accept: */*
>
* Request completely sent off
* schannel: remote party requests renegotiation
* schannel: renegotiating SSL/TLS connection
* schannel: SSL/TLS connection renegotiated
< HTTP/1.1 302 Found
< Location: https://localhost:59569/admin/
< Referrer-Policy: no-referrer
< Strict-Transport-Security: max-age=31536000; includeSubDomains
< X-Content-Type-Options: nosniff
< X-XSS-Protection: 1; mode=block
< content-length: 0
<
* Connection #0 to host localhost left intact
```

Because this is running on localhost, it's setting this policy for everything running on localhost, which is super undesirable.

**Reply by DamianEdwards** [reply]
Seems Keycloak defaults realms to sending HSTS response headers, even for localhost, see #14162. This is pretty dev-hostile behavior. HSTS headers should never be set for localhost responses.

**Reply by shawkins** [reply]
> @shawkins looking further at the headers coming back from the initial request to Keycloak, it seems like it's actually sending the HSTS header?

Yes, I missed that initially. At least it's clear what is occurring then. It would be better to log a new issue than to add comments to the old one - the concern about localhost or dev mode in general seems valid.

**Reply by DamianEdwards** [reply]
Thanks, logged #32366
